mlabs/lendex-endpoint-spec.md

```diff
@@ -0,0 +1,205 @@
+
+### Borrow
+
+Input:
+Mlabs.Lending.Contract.Api.Borrow
+(amount, asset, rate)
+ 
+
+Users: Borrowers
+
+Prerequisite:
+the user must have an appropriate amount of aTokens already held in the contract Collateral such that their borrow specifications can be met without putting them below minimum collateral Ratio
+
+collateral ratios are calculated using each aToken : underlying exchange rate, then the exchange rate between that underlying token and Ada, which is provided by the Oracle - this way the collateral and the borrowed currency can be compared via relative value in Ada.
+
+expected outcomes:
+- interest must be advanced on any outstanding borrows for this user - all calculations must use this amount and not the pre-interest amount.
+
+- wallet of user balance increases in `amount` of the Lendex's underlying token.
+- sufficient state must be maintained such that the user's outstanding borrow can always be known or accurately calculated, including interest.
+- if the user has existing borrows against their collateral, this must be considered as well such that the user's maximum borrow amount is never exceeded, interest should be recalculated before this check as well if needed.
+
+invariant behaviours:
+the user should not be able to specify negative amounts to borrow or provide collateral.  there should be no input such that this endpoint releases collateral to the user.
+
+Attack Vectors:
+
+Market Manipulation
+a user may attempt to manipulate the market or tamper with oracle behavior, being able to take out a loan during a price fluctuation (where the value of the underlying Token being borrowed is moved down, or the value of the collateral is move upward) allowing a borrow of more funds than the collateral provided once prices normalize
+
+This may require mitigation with time-weighted-averaging to the contract's benefit. 
+
+### Repay
+
+input:
+Mlabs.Lending.Contract.Api.Repay
+(amount, asset, rate)
+
+Users: Borrowers
+
+Prerequisite: 
+the user must have a loan, and in general will have Collateral to secure the loan up to the collateralRatio for the Lendex
+
+there are edge cases where we are unable to liquidate or have not yet liquidated the user, and they repay a portion of the loan, saving the loan from liquidation.
+
+expected outcomes
+interest against the loan should be recalculated.
+the wallet of the user will decrease by the specified amount from input in the Market's underlying Currency (or round down if the amount available is less than the amount in the input)
+the outstanding loan payment will be reduced by the amount paid, less interest applied against the loan
+Collateral is not released on this endpoint, even if the borrow amount is totally paid off.
+
+if repaying Ada, the user's stake delegation should be changed to some default? - if this feature is available in plutus
+
+invariant outcomes:
+can't repay a negative amount but 0 is ok.
+
+### SwapBorrowRateModel
```

Comment:
I guess it's not implemented yet. In Aave it allows to switch between Stable and Variable Rate models.
Those models differ in quotients of the transition function.

Reply:
can you describe what Stable and Variable rates mean? is this documented elsewhere?